### PR TITLE
Fix S107: 'Methods with too many parameters' should not raise on cons…

### LINKED ISSUE #1015: Fix S107: "Methods with too many parameters" should not raise on constructors calling base
### Description

This rule comes up frequently for constructors due to dependency injection scenarios, especially with ASP.NET Core / Identity framework related classes.

### Example

The [`Microsoft.AspNetCore.Identity.UserManager<TUser>`](https://github.com/aspnet/Identity/blob/dev/src/Microsoft.Extensions.Identity.Core/UserManager.cs#L66) class has a constructor with several parameters:

```
public UserManager(IUserStore<User> store, 
            IOptions<IdentityOptions> optionsAccessor,
            IPasswordHasher<User> passwordHasher, 
            IEnumerable<IUserValidator<User>> userValidators,
            IEnumerable<IPasswordValidator<User>> passwordValidators, 
            ILookupNormalizer keyNormalizer, IdentityErrorDescriber errors, 
            IServiceProvider services, 
            ILogger<UserManager<User>> logger) { }
```

There are other similar library classes which are often subclassed to override specific functionality.  Issues are generated by this rule in cases which are unavoidable, because while we might prefer to not have so many arguments, it's necessary to pass these along to base members.

### Expected behavior

I'm not sure what's ideal here.  Either it should be possible to ignore constructors entirely for this rule, or perhaps examine constructors for base constructor usage.  So for instance, if the constructor calls a base constructor, the identical parameters of the base constructor are excluded from the total parameter count.  This way you could check that too many additional parameters are not added.

I could see a similar issue with overridden methods, but I think this is far more common with constructors due to DI.

### Known workarounds

Not sure.  Is there a way to exclude code blocks from SonarC# coverage?

### Related information

SonarC# Version is  6.5 (build 3766)

### PR DESCRIPTION
**Pull request by Evangelink**
…tructors calling base

Fix #1015

### CODE REVIEW

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/TooManyParameters.cs [hunk 1]
```diff
@@ -31,4 +31,21 @@ public interface If
...
+        public MyWrongClass(string a, string b, string c, string d, string e, string f, string g, string h) // Noncompliant
+        {
+        }
+    }
+
+    public class SubClass : MyWrongClass
+    {
+        // See https://github.com/SonarSource/sonar-csharp/issues/1015
+        // We should not raise when parent base class forces usage of too many args
+        public SubClass(string a, string b, string c, string d, string e, string f, string g, string h) // Compliant (base class requires them)
```
Comment (duncanp-sonar) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/TooManyParameters.cs]: Could add a test case for where the base is compliant but the child isn't

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TooManyParameters.cs [hunk 2]
```diff
@@ -84,6 +84,14 @@ private static bool CanBeChanged(SyntaxNode node, SemanticModel semanticModel)
                 return true;
             }
 
+            if ((node as ConstructorDeclarationSyntax)?.Initializer?.ArgumentList?.Arguments.Count > Maximum)
+            {
+                // Base class is already not compliant so let's ignore current constructor.
+                // Another option could be to substract current number of parameters from base count and raise only if greater
+                // than threshold.
+                return false;
+            }
```
Comment (duncanp-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TooManyParameters.cs]: For consistency, should we not raise the issue on overridden methods too?

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/TooManyParameters.cs (hunk 1 above)
Comment (Evangelink) [reply sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/TooManyParameters.cs]: Will do

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TooManyParameters.cs (hunk 2 above)
Comment (Evangelink) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TooManyParameters.cs]: It is already handled (interface implementation + override) on the next lines (through semantic model).

**Pull request merged**